controlplane/kubeadm/internal/controllers/controller.go

```diff
@@ -791,6 +810,88 @@ func (r *KubeadmControlPlaneReconciler) reconcileEtcdMembers(ctx context.Context
 	return nil
 }
 
+func (r *KubeadmControlPlaneReconciler) reconcilePreTerminateHook(ctx context.Context, controlPlane *internal.ControlPlane) (ctrl.Result, error) {
+	if !controlPlane.HasDeletingMachine() {
+		return ctrl.Result{}, nil
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+
+	// Pick the Machine with the oldest deletionTimestamp to keep this function deterministic / reentrant
+	// so we only remove the pre-terminate hook from one Machine at a time.
+	deletingMachine := controlPlane.DeletingMachines().OldestDeletionTimestamp()
+	log = log.WithValues("Machine", klog.KObj(deletingMachine))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	// Return early if there are other pre-terminate hooks for the Machine.
+	// The KCP pre-terminate hook should be the one executed last, so that kubelet
+	// is still working while other pre-terminate hooks are run.
+	if machineHasOtherPreTerminateHooks(deletingMachine) {
+		return ctrl.Result{RequeueAfter: deleteRequeueAfter}, nil
+	}
+
+	// Return early if we don't have to run the pre-terminate hook (usually because it was already completed).
+	// We are now waiting for the Machine to go away.
+	if _, exists := deletingMachine.Annotations[controlplanev1.PreTerminateDeleteHookAnnotation]; !exists {
+		return ctrl.Result{RequeueAfter: deleteRequeueAfter}, nil
+	}
```

Comment:
Thought: Should we maybe instead (at the top) loop over all deleting machines and return early if **one** of them does not have the termination delete hook?

This might be a very very side use-case which normally should not happen, but maybe could (time-drift on kube-apiserver's?)

Reply:
Done